internal/services/authorization/role_assignment_resource_base.go

```diff
@@ -0,0 +1,383 @@
+package authorization
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-azure-helpers/lang/pointer"
+	"github.com/hashicorp/go-azure-helpers/lang/response"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonids"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/authorization/2022-04-01/roleassignments"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/authorization/2022-04-01/roledefinitions"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/azure"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/tf"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/authorization/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+type roleAssignmentBaseResource struct{}
+
+func (br roleAssignmentBaseResource) arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"principal_id": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
+
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.IsUUID,
+		},
+
+		"role_definition_id": {
+			Type:          pluginsdk.TypeString,
+			Optional:      true,
+			ForceNew:      true,
+			ConflictsWith: []string{"role_definition_name"},
+			ValidateFunc:  validation.StringIsNotEmpty,
+		},
+
+		"role_definition_name": {
+			Type:          pluginsdk.TypeString,
+			Optional:      true,
+			ForceNew:      true,
+			ConflictsWith: []string{"role_definition_id"},
+			ValidateFunc:  validation.StringIsNotEmpty,
+		},
+
+		"skip_service_principal_aad_check": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			ForceNew: true,
+			Default:  false,
+		},
+
+		"delegated_managed_identity_resource_id": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: azure.ValidateResourceID,
+		},
```

Comment:
Could we use `commonids.ValidateUserAssignedIdentityID` here?

Reply:
Probably no. According to this [issue](https://github.com/hashicorp/terraform-provider-azurerm/pull/11848#issue-900483150): this property does not mean the managed identity resource id, but the resource which contains a managed identity.